airbyte-integrations/connectors/source-paypal-transaction/source_paypal_transaction/source.py

```diff
@@ -0,0 +1,390 @@
+# SOFTWARE.
+#
+
+import time
+from abc import ABC
+from datetime import datetime, timedelta
+from typing import Any, Callable, Iterable, List, Mapping, MutableMapping, Optional, Tuple, Union
+
+import requests
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.http import HttpStream
+from airbyte_cdk.sources.streams.http.auth import HttpAuthenticator, Oauth2Authenticator
+from dateutil.parser import isoparse
+
+
+def get_endpoint(is_sandbox: bool = False) -> str:
+    if is_sandbox:
+        endpoint = "https://api-m.sandbox.paypal.com"
+    else:
+        endpoint = "https://api-m.paypal.com"
+    return endpoint
+
+
+class PaypalTransactionStream(HttpStream, ABC):
+
+    page_size = "500"  # API limit
+
+    # Date limits are needed to prevent API error: Data for the given start date is not available
+    # API limit: (now() - start_date_min) < start_date < (now() - start_date_max)
+    start_date_min: Mapping[str, int] = {"days": 3 * 364}  # API limit - 3 years
+    start_date_max: Mapping[str, int] = {"hours": 0}
+
+    stream_slice_period: Mapping[str, int] = {"days": 1}  # max period is 31 days (API limit)
+
+    requests_per_minute: int = 30  # API limit is 50 reqs/min from 1 IP to all endpoints, otherwise IP is banned for 5 mins
+
+    def __init__(
+        self,
+        authenticator: HttpAuthenticator,
+        start_date: Union[datetime, str],
+        end_date: Union[datetime, str] = None,
+        is_sandbox: bool = False,
+        **kwargs,
+    ):
+
+        self.start_date = start_date
+        self.end_date = end_date
+        self.is_sandbox = is_sandbox
+
+        # self._validate_input_dates()
+
+        super().__init__(authenticator=authenticator)
+
+    @property
+    def start_date(self) -> datetime:
+        return self._start_date
+
+    @start_date.setter
+    def start_date(self, value: Union[datetime, str]):
+        if isinstance(value, str):
+            value = isoparse(value)
+        self._start_date = value
+
+    @property
+    def minimum_allowed_start_date(self) -> datetime:
+        return self.now - timedelta(**self.start_date_min)
+
+    @property
+    def maximum_allowed_start_date(self) -> datetime:
+        return min(self.now - timedelta(**self.start_date_max), self.end_date)
+
+    @property
+    def end_date(self) -> datetime:
+        """Return initiated end_date or now()"""
+        if not self._end_date or self._end_date > self.now:
+            # If no end_date or end_date is in future then return now:
+            return self.now
+        else:
+            return self._end_date
+
+    @end_date.setter
+    def end_date(self, value: Union[datetime, str]):
+        if isinstance(value, str):
+            value = isoparse(value)
+        self._end_date = value
+
+    @property
+    def now(self) -> datetime:
+        return datetime.now().replace(microsecond=0).astimezone()
+
+    def _validate_input_dates(self):
```

Comment:
this method is not called anywhere

Reply:
it was called from SourcePaypalTransaction.check_connection() to validate dates in config